File: algorithms/03_cpp/numerical/least_squares_linear.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <functional>
#include <iostream>
#include <utility>
#include <vector>

std::pair<double, double> leastSquaresLinear(const std::vector<double> &xs, const std::vector<double> &ys) {
    double n = xs.size(), meanX = 0.0, meanY = 0.0;
    for (size_t i = 0; i < xs.size(); ++i) { meanX += xs[i] / n; meanY += ys[i] / n; }
    double numerator = 0.0, denominator = 0.0;
    for (size_t i = 0; i < xs.size(); ++i) {
        numerator += (xs[i] - meanX) * (ys[i] - meanY);
        denominator += (xs[i] - meanX) * (xs[i] - meanX);
    }
    double slope = numerator / denominator;
    return {meanY - slope * meanX, slope};
}

int main() {
    auto [intercept, slope] = leastSquaresLinear({0, 1, 2, 3}, {1, 3, 5, 7});
    assert(std::abs(intercept - 1.0) < 1e-12 && std::abs(slope - 2.0) < 1e-12);
    std::cout << "least squares linear ok\n";
    return 0;
}

File: algorithms/07_typescript/math/least_squares_linear.ts
// Ordinary least squares for a straight line.
function leastSquaresLinear(xs: number[], ys: number[]): { intercept: number; slope: number } {
  const n = xs.length;
  const meanX = xs.reduce((a, b) => a + b, 0) / n;
  const meanY = ys.reduce((a, b) => a + b, 0) / n;
  const numerator = xs.reduce((acc, x, i) => acc + (x - meanX) * (ys[i] - meanY), 0);
  const denominator = xs.reduce((acc, x) => acc + (x - meanX) ** 2, 0);
  const slope = numerator / denominator;
  return { intercept: meanY - slope * meanX, slope };
}

const { intercept, slope } = leastSquaresLinear([0, 1, 2, 3], [1, 3, 5, 7]);
if (Math.abs(intercept - 1) > 1e-12 || Math.abs(slope - 2) > 1e-12) throw new Error("least squares failed");
console.log(`y=${intercept}+${slope}x`);
